grails-app/services/org/pih/warehouse/data/ProductAssociationDataService.groovy

```diff
@@ -0,0 +1,141 @@
+            productAssociationInstance.associatedProduct = Product.findByProductCode(params['associatedProduct.productCode'])
+            params['product.id'] =  productAssociationInstance.product?.id
+            params['associatedProduct.id'] =  productAssociationInstance.associatedProduct?.id
+
+            // disable update product association on import
+            if (productAssociationInstance?.id) {
+                command.errors.reject("Row ${index + 1}: Cannot edit existing associations via import")
+            }
+            if (!productAssociationInstance.product) {
+                command.errors.reject(
+                    "Row ${index + 1}: Product with code '${params['product.productCode']}' does not exist"
+                )
+            }
+            if (!productAssociationInstance.associatedProduct) {
+                command.errors.reject(
+                    "Row ${index + 1}: Product with code '${params['associatedProduct.productCode']}' does not exist"
+                )
+            }
+            if (productAssociationInstance.product && !productAssociationInstance.product.active) {
+                command.errors.reject(
+                    "Row ${index + 1}: Product with code '${productAssociationInstance.product?.productCode}' is inactive"
+                )
+            }
+            if (productAssociationInstance.associatedProduct && !productAssociationInstance.associatedProduct.active) {
+                command.errors.reject(
+                    "Row ${index + 1}: Product with code '${productAssociationInstance.associatedProduct?.productCode}' is inactive"
+                )
+            }
+            if (productAssociationInstance.product?.id == productAssociationInstance.associatedProduct?.id) {
+                command.errors.reject("Cannot associate a product with itself")
+            }
+
+            // Check for association duplicates
+            List<ProductAssociation> foundProductAssociations = ProductAssociation.findAllWhere([
+                    product             : productAssociationInstance.product,
+                    associatedProduct   : productAssociationInstance.associatedProduct,
+                    code                : productAssociationInstance.code,
+            ])
+            if (foundProductAssociations && foundProductAssociations.size() > 0) {
+                command.errors.reject("Row ${index + 1}: Association already exists")
+            }
+
+
+            // find another product association that matches current associationInstance
+            // for two-way association relationship, where:
+            // - association.product = otherAssociation.associatedProduct
+            // - association.associatedProduct = otherAssociation.product
+            // - association.code = otherAssociation.code
+            def otherMatchingTwoWayAssociation = command.data.find {
```

Comment:
Should not this be searched in all associations in the system instead in the file data only?

Reply:
I think this might be a topic for a tech huddle and at the moment we should focus only on associations provided in the imported file.

This is also falls into a case of updating associations which is probably out of scope for this ticket